storage/innobase/handler/ha_innodb.cc

```diff
@@ -7730,26 +7730,56 @@ ha_innobase::build_template(
 
 	ulint num_v = 0;
 
-	if (active_index != MAX_KEY
-	     && active_index == pushed_idx_cond_keyno) {
-		m_prebuilt->idx_cond = this;
-		goto icp;
-	} else if (pushed_rowid_filter && rowid_filter_is_active) {
-icp:
-		/* Push down an index condition or an end_range check. */
+	/* MDEV-31154: For pushed down index condition we don't support virtual
+	column and idx_cond_push() does check for it. For row ID filtering we
+	don't need such restrictions but we get into trouble trying to use the
+	ICP path.
+
+	1. It should be fine to follow no_icp path if primary key is generated.
+	However, with user specified primary key(PK), the row is identified by
+	the PK and those columns need to be converted to mysql format in
+	row_search_idx_cond_check before doing the comparison. Since secondary
+	indexes always have PK appended in innodb, it works with current ICP
+	handling code when fetch_primary_key_cols is set to TRUE.
+
+	2. Although ICP comparison and Row ID comparison works on different
+	columns the current ICP code can be shared by both.
+
+	3. In most cases, it works today by jumping to goto no_icp when we
+	encounter a virtual column. This is hackish and already have some
+	issues as it cannot handle PK and all states are not reset properly,
+	for example, idx_cond_n_cols is not reset.
+
+	4. We already encountered MDEV-28747 m_prebuilt->idx_cond was being set.
+
+	Neither ICP nor row ID comparison needs virtual columns and the code is
+	simplified to handle both. It should handle the issues. */
+
+	const bool pushed_down = active_index != MAX_KEY
+				 && active_index == pushed_idx_cond_keyno;
+	const bool rowid_filter = pushed_rowid_filter && rowid_filter_is_active;
+
+	m_prebuilt->idx_cond = pushed_down ? this : nullptr;
+
+	if (pushed_down || rowid_filter) {
```

Comment:
The compiler might be smart enough to defer the evaluation of `rowid_filter`, but we might also avoid introducing that variable:
```c++
	m_prebuilt->idx_cond = pushed_down ? this : nullptr;

	if (pushed_down || (pushed_rowid_filter && rowid_filter_is_active)) {
```
Even better, I see that `pushed_rowid_filter && rowid_filter_is_active` was already evaluated earlier in this function. `rowid_filter` is actually duplicating an existing predicate:
```c++
	if (pushed_down || m_prebuilt->pk_filter) {
```

Reply:
In general I would prefer code clarity and readability over such local optimizations unless in true performance critical path and we know the reason precisely. 

m_prebuilt->pk_filter is a good observation and I think we could replace the variable with it. To keep it consistent I have put the condition as 
if (m_prebuilt->idx_cond  || m_prebuilt->pk_filter)